### PR TITLE
implement CDB cache versioning via timestamp (closes #714)

### LINKED ISSUE #714: CDB cache
It looks like the engine file system doesn't flush the files immediately to disk. If the engine will be closed before the cache is flushed to disk, the cache files will be broken.

CDB cache was added in #698.

- [x] Flush to disk immediately (ad881d6071c5f8b7e48e11db3086d6cad799af92)
	- [x] ~~Sub-task: Rescan CDB cache folder (#709)~~
		- [x] Rescan not needed, it was a bug fixed in ad881d6071c5f8b7e48e11db3086d6cad799af92
- [x] Version check (#727, #762)
- [x] Integrity check (a0f045012700ffeed8ee07d9ad0ca6ab9a6c7e7c)

### PR DESCRIPTION
**Pull request by jjdredd**
результаты легкого тестирования
- старый кеш не загружается т.к. у него нет версии, генерится новый
- новый кеш загружается при перезагрузке сейва не выходя из игры и после полного перезапуска
- если удалить кеш то сгенерится новый который загружается при полном перезапуске

### CODE REVIEW

**Comment by jjdredd** [comment]
:yyy:

**Comment by vTurbine** [comment]
Did you check how much time checksum calculation takes in each case? Don't know much about internals of XRay FS but IO can introduce some unwanted latencies here. Wouldn't be better to check, let say, file modification date only?

**Comment by jjdredd** [comment]
Да, мне тоже как-то показалось что это оверкил, но тут настаивали на таком решении. может есть причина.

**Comment by Xottab-DUTY** [comment]
Please, keep conversation in English.

I didn't insisted that hard, though. You just needed details on how to implement it and I told you the first idea I had in my head.

**Comment by jjdredd** [comment]
> Please, keep conversation in English.
> I _didn't insisted_ that hard, though.
 
I'm not sure why, but if you require me to use English, make effort to use proper English grammar yourself.
I don't know, it didn't sound like a suggestion to me. If there's a general consensus on how this feature 
must be implemented, I can send a new patch.

Edit: Although I'd hate to redo everything, I like @vTurbine's method better.

**Comment by jjdredd** [comment]
Also, wouldn't it better to just check file modification date for shaders too?

**Comment by Xottab-DUTY** [comment]
> Edit: Although I'd hate to redo everything, I like @vTurbine's method better.

I think you don't need to redo it, just replace crc32 with date :D



> Also, wouldn't it better to just check file modification date for shaders too?

No, shaders are text files, so it's worth to read them.

**Comment by jjdredd** [comment]
On a second thought: all the functions that build caches accept an IReader, not a file name, so the particular file name is not accessible.  In order to check modification time of a file I'd have to change the prototypes of those functions which is not good.

File: src/Layers/xrRenderPC_R1/FStaticRender_Loader.cpp [hunk 1]
```diff
@@ -26,7 +26,7 @@ void CRender::level_Load(IReader* fs)
     g_pGamePersistent->LoadTitle();
     {
         chunk = fs->open_chunk(fsL_SHADERS);
-        R_ASSERT2(chunk, "Level doesn't builded correctly.");
+        R_ASSERT2(chunk, "Level isn't built correctly.");
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_R1/FStaticRender_Loader.cpp]: Don't do unrelated changes.
1. Since this is unrelated to this PR (which is dedicated only to CDB cache), the grammar can be fixed in separate PR.
2. But I think that original GSC text is canonical, so let's not touch it.
```suggestion
        R_ASSERT2(chunk, "Level doesn't builded correctly.");
```

File: src/xrCDB/xrCDB.h [hunk 2]
```diff
@@ -91,6 +92,7 @@ class XRCDB_API MODEL : Noncopyable
     void build(Fvector* V, int Vcnt, TRI* T, int Tcnt, build_callback* bc = NULL, void* bcp = NULL);
     u32 memory();
 
+    void set_version(u32 version) { this->version = version; }
```
Comment (Xottab-DUTY) [inline src/xrCDB/xrCDB.h]:
```suggestion
    void set_version(u32 value) { version = value; }
```

File: src/xrCDB/xr_area.cpp [hunk 3]
```diff
@@ -106,33 +106,34 @@ void CObjectSpace::Load(IReader* F, CDB::build_callback build_callback)
...
     Fvector* verts = (Fvector*)F->pointer();
     CDB::TRI* tris = (CDB::TRI*)(verts + H.vertcount);
+    Static.set_version(F->get_age());
     Create(verts, tris, H, build_callback);
     FS.r_close(F);
 }
 
-void CObjectSpace::Create(Fvector* verts, CDB::TRI* tris, const hdrCFORM& H, CDB::build_callback build_callback)
+void CObjectSpace::Create(Fvector* verts, CDB::TRI* tris, const hdrCFORM& H,
+                          CDB::build_callback build_callback)
```
Comment (Xottab-DUTY) [inline src/xrCDB/xr_area.cpp]: No unrelated formatting changes needed.
```suggestion
void CObjectSpace::Create(Fvector* verts, CDB::TRI* tris, const hdrCFORM& H, CDB::build_callback build_callback)
```

File: src/xrCore/FS.h [hunk 4]
```diff
@@ -330,6 +332,7 @@ class IReaderBase
 
 private:
     size_t m_last_pos;
+    u32 file_age;
```
Comment (Xottab-DUTY) [inline src/xrCore/FS.h]:
```suggestion
    u32 m_file_age;
```

File: src/xrCore/FS.h [hunk 5]
```diff
@@ -192,11 +192,13 @@ class IReaderBase
...
 {
 public:
-    IC IReaderBase() : m_last_pos(0) {}
+    IC IReaderBase() : m_last_pos(0), file_age(0) {}
     virtual ~IReaderBase() = default;
     IC implementation_type& impl() { return *(implementation_type*)this; }
     IC const implementation_type& impl() const { return *(implementation_type*)this; }
 
+    IC void set_age(u32 age) { file_age = age; }
+    IC u32 get_age() { return file_age; }
```
Comment (Xottab-DUTY) [inline src/xrCore/FS.h]:
```suggestion
    IC void set_age(u32 age) { m_file_age = age; }
    IC u32 get_age() const { return m_file_age; }
```

File: src/Layers/xrRender/HOM.cpp [hunk 6]
```diff
@@ -124,23 +124,23 @@ void CHOM::Load()
...
+    m_pModel->set_version(fs->get_age());
+    bool bUseCache = strstr(Core.Params, "-cdb_cache");
+
+    strconcat(fName, "cdb_cache" DELIMITER, FS.get_path("$level$")->m_Add, "hom.bin");
+    FS.update_path(fName, "$app_data_root$", fName);
+
+    if ( bUseCache && FS.exist(fName) && m_pModel->deserialize(fName) )
+    {
+        Log("* Loaded HOM cache...");
+    }
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/HOM.cpp]:
```suggestion
    bool bUseCache = strstr(Core.Params, "-cdb_cache");

    strconcat(fName, "cdb_cache" DELIMITER, FS.get_path("$level$")->m_Add, "hom.bin");
    FS.update_path(fName, "$app_data_root$", fName);

    if (bUseCache && FS.exist(fName) && m_pModel->deserialize(fName))
    {
#ifndef MASTER_GOLD
        Msg("* Loaded HOM cache (%s)...", fName);
#endif
    }
```

File: src/Layers/xrRender/HOM.cpp [hunk 7]
```diff
@@ -124,23 +124,23 @@ void CHOM::Load()
...
     else
     {
-        strconcat(fName, "cdb_cache" DELIMITER, FS.get_path("$level$")->m_Add, "hom.bin");
-        FS.update_path(fName, "$app_data_root$", fName);
-        if (!FS.exist(fName))
-        {
-            Msg("* HOM cache for '%s' not found. Building the model from scratch..", fName);
-            m_pModel->build(CL.getV(), int(CL.getVS()), CL.getT(), int(CL.getTS()));
+        Msg("* HOM cache for '%s' was not loaded. Building the model from scratch..", fName);
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/HOM.cpp:142]:
```suggestion
#ifndef MASTER_GOLD
        Msg("* HOM cache for '%s' was not loaded. Building the model from scratch..", fName);
#endif
```

File: src/Layers/xrRenderPC_R1/FStaticRender_Loader.cpp [hunk 8]
```diff
@@ -360,15 +360,16 @@ void CRender::LoadSectors(IReader* fs)
...
+        if (use_cache && FS.exist(fName) && rmPortals->deserialize(fName))
         {
-            if (!FS.exist(fName))
-                Msg("* Portals cache for '%s' not found. Building the model from scratch..", fName);
-            else
-            {
-                rmPortals->deserialize(fName);
-                do_rebuild = false;
-            }
+            Log("* Loaded portals cache...");
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_R1/FStaticRender_Loader.cpp]:
```suggestion
#ifndef MASTER_GOLD
            Msg("* Loaded portals cache (%s)...", fName);
#endif
```

File: src/Layers/xrRenderPC_R1/FStaticRender_Loader.cpp [hunk 9]
```diff
@@ -360,15 +360,16 @@ void CRender::LoadSectors(IReader* fs)
...
+            do_rebuild = false;
+        }
+        else
+        {
+            Msg("* Portals cache for '%s' was not loaded. "
+                "Building the model from scratch..", fName);
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_R1/FStaticRender_Loader.cpp:375]:
```suggestion
#ifndef MASTER_GOLD
            Msg("* Portals cache for '%s' was not loaded. "
                "Building the model from scratch..", fName);
#endif
```

File: src/Layers/xrRender_R2/r2_loader.cpp [hunk 10]
```diff
@@ -355,16 +355,16 @@ void CRender::LoadSectors(IReader* fs)
...
-                rmPortals->deserialize(fName);
-                do_rebuild = false;
-            }
+            Log("* Loaded portals cache...");
+            do_rebuild = false;
+        }
+        else
+        {
+            Msg("* Portals cache for '%s' was not loaded. "
+                "Building the model from scratch..", fName);
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender_R2/r2_loader.cpp:370]:
```suggestion
#ifndef MASTER_GOLD
            Msg("* Portals cache for '%s' was not loaded. "
                "Building the model from scratch..", fName);
#endif
```

File: src/Layers/xrRender_R2/r2_loader.cpp [hunk 11]
```diff
@@ -355,16 +355,16 @@ void CRender::LoadSectors(IReader* fs)
...
         {
-            if (!FS.exist(fName))
-                Msg("* Portals cache for '%s' not found. Building the model from scratch..", fName);
-            else
-            {
-                Log("* Loading portals cache...");
-                rmPortals->deserialize(fName);
-                do_rebuild = false;
-            }
+            Log("* Loaded portals cache...");
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender_R2/r2_loader.cpp]:
```suggestion
#ifndef MASTER_GOLD
            Msg("* Loaded portals cache (%s)...", fName);
#endif
```

**Review by Xottab-DUTY — CHANGES_REQUESTED** [review]

**Review by Xottab-DUTY — APPROVED** [review]
